src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs

```diff
@@ -58,42 +64,88 @@ private unsafe static bool TryParseSlow(ref BufferReader<byte> reader, out bool
         /// Try to parse an int out of the reader (as UTF-8).
         /// </summary>
         /// <param name="value">The parsed int value or default if failed.</param>
+        /// <param name="standardFormat">Expected format.</param>
+        /// <remarks>
+        /// <see cref="Utf8Parser.TryParse(ReadOnlySpan{byte}, out int, out int, char)"/> for supported formats.
+        /// </remarks>
+        /// <exception cref="FormatException">Thrown if the given format isn't supported.</exception>
         /// <returns>True if successfully parsed.</returns>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static unsafe bool TryParse(ref this BufferReader<byte> reader, out int value)
+        public static unsafe bool TryParse(ref this BufferReader<byte> reader, out int value, char standardFormat = '\0')
         {
             ReadOnlySpan<byte> unread = reader.UnreadSpan;
-            if (Utf8Parser.TryParse(unread, out value, out int consumed) && consumed < unread.Length)
+            if (!Utf8Parser.TryParse(unread, out value, out int consumed, standardFormat))
+            {
+                return false;
+            }
+
+            if (consumed < unread.Length)
             {
+                // The parser found a digit it wouldn't consume so we have all valid data.
                 reader.Advance(consumed);
                 return true;
             }
 
-            return TryParseSlow(ref reader, out value);
+            // If we ate all of our unread there may be more valid digits
+            return TryParseSlow(ref reader, out value, standardFormat);
```

Comment:
This means we will always hit the slow path at the end of the sequence/span, i.e consumed will be equal to unread.Length. Can we avoid that?

Reply:
Not easy to do as we would need to know if we're in the last span. We could add a check for single segment, but the expectation is that you won't hit this often and we want to keep the fast path as small as possible.

Reply:
Maybe the check could be inside Slow method but before we actually allocate the buffer and copy?

Reply:
That we're single segment? We'd still need to parse again- and if we're truly single segment there will be no copy- peek will just slice.

Reply:
Ah, I see. I forgot that peek with simply return the last span (if it's single segment)